A29-xds-tls-security.md

```diff
@@ -0,0 +1,543 @@
+
+A secure client at a minimum requires [`validation_context_certificate_provider_instance`][VCCPI1]
+inside [`combined_validation_context`][CVC] to be able to validate the server certificate
+in the TLS mode. As part of validating all CDS updates, if this field is not set in a CDS,
+gRPC will NACK the CDS update. The client identity certificate is configured via the
+[`tls_certificate_certificate_provider_instance`][TCCPI] field which is optional. The
+client certificate is sent to the server if the server requests or requires it in the
+TLS handshake. If the server requires the client certificate but is not
+configured on the client side then the TLS handshake will fail because of the client's inability
+to send the certificate.
+
+As part of validating all CDS updates, if any of the following fields are present, gRPC
+will NACK a CDS update:
+
+* [sni][SNI]
+* [allow_renegotiation][ALLOW-RENEG]
+* [max_session_keys][MAX-SESS-KEYS]
+
+[A27:CDS]: A27-xds-global-load-balancing.md#cds
+[CL-TS]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L937
+[CL-TS-comment]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L932
+[CL-TS-matches]: https://github.com/envoyproxy/envoy/blob/9aca65c395f01020080166e4795455addde167fa/api/envoy/config/cluster/v3/cluster.proto#L680
+[CTC]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L37
+[SNI]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L40
+[ALLOW-RENEG]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L47
+[MAX-SESS-KEYS]: https://github.com/envoyproxy/envoy/blob/7d4b2cae486b66b62ba0d3e1e348504699bea1bf/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L53
+
+##### Server Authorization aka Subject-Alt-Name Checks
+
+"Server authorization" is performed by the client where the client authorizes
+a server for the connection as described below. This check is in lieu of the
+canonical [hostname check in Web PKI][RFC6125-6].
+
+If [`match_subject_alt_names`][match_subject_alt_names] in
+[`default_validation_context`][default_validation_context] of
+[combined_validation_context][CVC] of the received `UpstreamTlsContext` is populated
+then gRPC checks the SAN entries in the leaf server certificate against the
+`match_subject_alt_names` values as follows:
+
+* if the `match_subject_alt_names` list is empty then there is nothing to match and the
+check succeeds.
+
+* if there is no leaf server certificate, or there are no SAN entries in the certificate
+the check fails.
+
+* each SAN entry of type DNS, URI and IP address is considered for the below match
```

Comment:
Envoy also supports RFC 822 (email) names. Are we purposefully not supporting them?
https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/source/extensions/transport_sockets/tls/utility.cc#L174

Reply:
Envoy being a general purpose proxy and supporting other protocols (http1 etc) I thought they see more use cases for email names etc. I am not aware of use-cases or usage of email names in gRPC so I thought it will be okay to omit them.

Reply:
As discussed separately, I wasn't saying we _should_ support it but think we should either support it or call out that we are purposefully not supporting (or requiring support for) this. It seems very easy to support, so I'd just assume to support it, but I accept it isn't much of a mainstream feature and if we didn't support it we may never hear from a user noticing it isn't available.

Reply:
I think it's almost identical to URI matching so I don't see any reason to exclude support since the implementation cost is almost zero. Will update the text.

Reply:
Added support for email type